pkg/kubelet/certificate/kubelet_test.go

```diff
@@ -100,3 +104,199 @@ func TestAddressesToHostnamesAndIPs(t *testing.T) {
+APtq+UjevEIjva5Qge+Kj5X/zPcJq9HZV98uVSPMOWSC7GLTkjdS6IwjxeFivU8x
+2BROZ6N+8btxlv8JlZMYHhIRmctM5nPI8bVC12qv4HXQ0wIEbrTgnoc+NiH6gtGY
+n8eN5Tw33FUpkEhqinjR
+-----END CERTIFICATE-----
+`
+
+const key2 = `
+-----BEGIN PRIVATE KEY-----
+MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQDLPd24HHWDKHOP
+DoSooyRHdUq5WEE3dWcZm/oH/EBMy5Zb3nR6Fh+wF+2qopyjujQnx3lb8OjuTO7i
+BW1iqH3Nrqml7z5d4i4+/zSf0gh3BiWDyAKDidtwhfQi+wuXCAtphDXEdoAw5XOy
+dWmLJyyUSQJekN53+9M0dw5lVig1HTxggF7q1k4aCL/S9NcvqFeIFXhXYmpXEvSs
+CdpF8EP2DtA7Eq/I1PjB5BcHoSToY/bWKfEKBAzatzOR28t6Bp4VPEskXUFDeGwo
+vy2KWlOoIplEyNaX0srQq8fIoRG3dttbmnR0ZG6d1Vo0lrru7STs66iBUOtEJUHZ
+Vj7IWb3ZAgMBAAECggEBAKt1La88kwZrAdIV9WQu/VQrZzaldZ9LtAaux2glLjmb
+JuWp7alxMJpmFWJ2fJ7DX2yPo5okytz+miijW3x3mGoEh6otAvhA77LFqaeKkQmY
+bd2Wxkgh7LYoy6UXFNf9OWNy3ck4Dz0w1UIgO0HhcoJGdXFB1exyzeLc7ZAf9xuW
+ZZK/az4OmqtDd8vXElOkD6b56BbCtGVq4br4P533yesH4DDQpy7rkvCgHdro3wIH
+aeg4AE8LOMmZggBOLorKXvG3+9s5ZXEnHj00oAk0c27/4UgGeMI0DGAayRJfTLW7
+fZZRerVuGRC0KmLYjBT4Vj0Ih/2LSE1Y+ZE0ZUtTQmECgYEA5TtVTfHQWb3V+1N+
+Rxl0O7rg4O0PCKWDjnbqYLbBNKZIrar4ebfqSAYWuhXR6jasRDnBlMH5pt5LD0g4
+bcztKNSYajUaRBjNg6lHsvAp1AKgbLhpUtwXvNe4bjNJFM7yMY6o/ZhFVWwylLQI
+BX9RMWLF7bBxSx6QFko95eZz7P8CgYEA4vmVUC3DWVQ+KZb9hEtw4UKyWs+WMQTy
+KIE8ZJoAweFEqCvvk03DwAy5xlDwtOuyO/dDibMmoB7Jgkr4iB9MA0qJEjkjDihG
+mabUW2K3WTezEZT31EwjAIs8If7zKatSwHZOLTmAam70LSbAy4ZVL3yFf4UbT5Oh
+NsXQ77hoXScCgYBfkaQX8ff2YjHjLUUZaWBPQrNcsxiwdyjo7WT721WjmKv2U7By
+Np4jVv6EqHIy3oZlj1rIpTJrQoQyo9560JQTkMbWiLshpuGPwbSVwpD9xfaSPTQU
+CpSO87T9pL4UQc8xoBOOXryRR6Gy43fwqsrz9wUj+orRUbWqxVsXDURJiwKBgQCa
+VsSlZLj1QUeT2ExDbVkwk73b6lRiuM5BpL+AWQgyzg91m4qpS7PUH9Mje15yZ+Mm
+y5htRhj5wHWd14TwavexNTnH3npr4g8/5CV6jsHGNQ3a4sUy4yLZ99PH+ik3KHx4
+yvmV3wfnV9NJ8JQg0ROT2sScVdKgZe615AWTPH4a+QKBgDA7kyhEYoCltchrEsxk
+ahz4YVmnJVZp1UDy+bLUZYCKNCyGIAkTVNutrxHThx/3HapMjkd6EmefLyNZ6rd9
+iNByAPzy9m6/zVhMdy97RgDpbXR+iuHpCwjiTh7bF3l78fLgStB2UD9jgiNDaiBv
+KSUELgGVOfUpsLO/ZTQqdJEy
+-----END PRIVATE KEY-----
+`
+
+func TestKubeletServerCertificateFromFiles(t *testing.T) {
+	certDir, err := os.MkdirTemp("", "test-kubelet-cert-files")
+	if err != nil {
+		t.Fatalf("Unable to setup cert dir: %v", err)
+	}
+
+	rotateCertErrs := make(chan error, 10)
+
+	defer func() {
+		if err := os.RemoveAll(certDir); err != nil {
+			t.Errorf("Unable to clean up test directory %q: %v", certDir, err)
+		}
+		if len(rotateCertErrs) != 0 {
+			t.Errorf("got errors when rotating certificate files in the test")
+		}
+		close(rotateCertErrs)
+	}()
+
+	certPath := filepath.Join(certDir, "kubelet.cert")
+	keyPath := filepath.Join(certDir, "kubelet.key")
+	createCertFn := func(cert, key []byte) error {
+		if err := os.WriteFile(certPath, cert, os.FileMode(0644)); err != nil {
+			return err
+		}
+		if err := os.WriteFile(keyPath, key, os.FileMode(0600)); err != nil {
+			return err
+		}
+		return nil
+	}
+	err = createCertFn([]byte(cert1), []byte(key1))
+	if err != nil {
+		t.Fatalf("Unable to setup cert: %v", err)
+	}
+
+	// simulate certificate files update in the background
```

Comment:
The "correct" way to update files on disk is via atomic operations like move.  Probably worth testing both correct and incorrect approaches.

Reply:
yeah, we did use the atomic rename approach when updating such files.

I added the tests for both ways of updating files.  

-----

The underlying [DynamicCertKeyPairContent ](https://pkg.go.dev/k8s.io/apiserver/pkg/server/dynamiccertificates#DynamicCertKeyPairContent) handles various cert/key files changes very reliably.